sphinx/registry.py

```diff
@@ -153,11 +153,20 @@ def preload_builder(self, app: "Sphinx", name: str) -> None:
 
             self.load_extension(app, entry_point.module)
 
-    def create_builder(self, app: "Sphinx", name: str) -> Builder:
+    def create_builder(self, app: "Sphinx", name: str,
+                       env: BuildEnvironment = None) -> Builder:
         if name not in self.builders:
             raise SphinxError(__('Builder name %s not registered') % name)
 
-        return self.builders[name](app)
+        try:
+            return self.builders[name](app, env)
+        except TypeError:
+            warnings.warn(
+                f"The custom builder {name} defines a custom __init__ method without the "
+                f"'env'argument. Report this bug to the developers of your custom builder, "
+                f"this is likely not a issue with Sphinx. The 'env' argument will be required "
+                f"from Sphinx 7.", RemovedInSphinx70Warning, stacklevel=2)
+            return self.builders[name](app)
```

Comment:
I was concerned the potential to call `set_environment` twice, but it seems that wouldn't be a problem, so updated. I also added a check so that we only issue one warning.

> Please be careful to changing the API. I believe Sphinx is not only an application but also a framework. So I consider this is a mix of refactoring and API changing. Is it intended?

This PR shouldn't have any end-user facing changes. I would say I consider Sphinx more of an application with a plugin mechanism though, rather than a library/framework -- we should be able to make changes that improve things on our end, although of course seeking to keep backwards compatability.

A